numba/tests/support.py

```diff
@@ -409,27 +411,63 @@ def tweak_code(func, codestring=None, consts=None):
                       co.co_lnotab)
     func.__code__ = new_code
 
-def static_temp_directory(dirname):
-    """
-    Create a directory in the temp dir with a given name. Statically-named
-    temp dirs created using this function are needed because we can't delete a
-    DLL under Windows (this is a bit fragile if stale files can influence the
-    result of future test runs).
-    """
-    if os.name == 'nt':
-        # Under Windows, gettempdir() points to the user-local temp dir
-        tmpdir = os.path.join(tempfile.gettempdir(), dirname)
-    else:
-        # Mix the UID into the directory name to allow different users to
-        # run the test suite without permission errors (issue #1586)
-        tmpdir = os.path.join(tempfile.gettempdir(),
-                              "%s.%s" % (dirname, os.getuid()))
+
+_trashcan_dir = 'numba-tests'
+
+if os.name == 'nt':
+    # Under Windows, gettempdir() points to the user-local temp dir
+    _trashcan_dir = os.path.join(tempfile.gettempdir(), _trashcan_dir)
+else:
+    # Mix the UID into the directory name to allow different users to
+    # run the test suite without permission errors (issue #1586)
+    _trashcan_dir = os.path.join(tempfile.gettempdir(),
+                                 "%s.%s" % (_trashcan_dir, os.getuid()))
+
+# Stale temporary directories are deleted after they are older than this value.
+# The test suite probably won't ever take longer than this...
+_trashcan_timeout = 24 * 3600
+
+def _create_trashcan_dir():
     try:
-        os.mkdir(tmpdir)
+        os.mkdir(_trashcan_dir)
     except OSError as e:
         if e.errno != errno.EEXIST:
             raise
-    return tmpdir
+
+def _purge_trashcan_dir():
+    freshness_threshold = time.time() - _trashcan_timeout
+    for fn in sorted(os.listdir(_trashcan_dir)):
+        fn = os.path.join(_trashcan_dir, fn)
+        try:
+            st = os.stat(fn)
+            if st.st_mtime < freshness_threshold:
+                shutil.rmtree(fn, ignore_errors=True)
+        except OSError as e:
+            # In parallel testing, several processes can attempt to
+            # remove the same entry at once, ignore.
+            if e.errno != errno.ENOENT:
+                raise
+
+def _create_trashcan_subdir(prefix):
+    _purge_trashcan_dir()
+    path = tempfile.mkdtemp(prefix=prefix + '-', dir=_trashcan_dir)
+    return path
+
+def temp_directory(prefix):
+    """
+    Create a temporary directory with the given *prefix* that will survive
+    at least as long as this process invocation.  The temporary directory
+    will be eventually deleted when it becomes stale enough.
+
+    This is necessary because a DLL file can't be deleted while in use
+    under Windows.
+
+    An interesting side-effect is to be able to inspect the test files
+    shortly after a test suite run.
+    """
```

Comment:
Thanks for that rationale. I would still prefer if by default test artifacts (such as temporary directories) were cleaned up after each run, at least on sane platforms that allow that. That being said, I agree to the usefulness to be able to keep such data around for forensic analysis, but I would prefer that to be done through a command-line option (--keep-temporaries ?).